compat.c

```diff
@@ -176,59 +657,1456 @@ ssize_t iio_scan_context_get_info_list(struct iio_scan_context *ctx,
+	FIND_SYMBOL(ctx->lib, iio_channel_attr_get_filename);
+	FIND_SYMBOL(ctx->lib, iio_channel_attr_read_raw);
+	FIND_SYMBOL(ctx->lib, iio_channel_attr_read_bool);
+	FIND_SYMBOL(ctx->lib, iio_channel_attr_read_longlong);
+	FIND_SYMBOL(ctx->lib, iio_channel_attr_read_double);
+	FIND_SYMBOL(ctx->lib, iio_channel_attr_write_raw);
+	FIND_SYMBOL(ctx->lib, iio_channel_attr_write_string);
+	FIND_SYMBOL(ctx->lib, iio_channel_attr_write_bool);
+	FIND_SYMBOL(ctx->lib, iio_channel_attr_write_longlong);
+	FIND_SYMBOL(ctx->lib, iio_channel_attr_write_double);
+	FIND_SYMBOL(ctx->lib, iio_channel_get_type);
+	FIND_SYMBOL(ctx->lib, iio_channel_get_modifier);
+	FIND_SYMBOL(ctx->lib, iio_channel_get_index);
+	FIND_SYMBOL(ctx->lib, iio_channel_get_data_format);
+	FIND_SYMBOL(ctx->lib, iio_channel_convert);
+	FIND_SYMBOL(ctx->lib, iio_channel_convert_inverse);
+	FIND_SYMBOL(ctx->lib, iio_channel_enable);
+	FIND_SYMBOL(ctx->lib, iio_channel_disable);
+	FIND_SYMBOL(ctx->lib, iio_channel_is_enabled);
+
+	FIND_SYMBOL(ctx->lib, iio_create_channels_mask);
+	FIND_SYMBOL(ctx->lib, iio_channels_mask_destroy);
+
+	FIND_SYMBOL(ctx->lib, iio_device_create_buffer);
+	FIND_SYMBOL(ctx->lib, iio_buffer_destroy);
+	FIND_SYMBOL(ctx->lib, iio_buffer_cancel);
+	FIND_SYMBOL(ctx->lib, iio_buffer_enable);
+	FIND_SYMBOL(ctx->lib, iio_buffer_disable);
+	FIND_SYMBOL(ctx->lib, iio_buffer_get_device);
+	FIND_SYMBOL(ctx->lib, iio_buffer_get_channels_mask);
+	FIND_SYMBOL(ctx->lib, iio_buffer_get_data);
+	FIND_SYMBOL(ctx->lib, iio_buffer_set_data);
+
+	FIND_SYMBOL(ctx->lib, iio_buffer_create_block);
+	FIND_SYMBOL(ctx->lib, iio_block_destroy);
+	FIND_SYMBOL(ctx->lib, iio_block_get_buffer);
+	FIND_SYMBOL(ctx->lib, iio_block_enqueue);
+	FIND_SYMBOL(ctx->lib, iio_block_dequeue);
+	FIND_SYMBOL(ctx->lib, iio_block_start);
+	FIND_SYMBOL(ctx->lib, iio_block_first);
+	FIND_SYMBOL(ctx->lib, iio_block_end);
+	FIND_SYMBOL(ctx->lib, iio_block_foreach_sample);
+
+	FIND_SYMBOL(ctx->lib, iio_strlcpy);
+	FIND_SYMBOL(ctx->lib, iio_strdup);
+	FIND_SYMBOL(ctx->lib, iio_strerror);
+	FIND_SYMBOL(ctx->lib, iio_prm_printf);
+
+#undef FIND_SYMBOL
+
+	compat_lib_ptr = ctx;
+
+	IIO_CALL(iio_prm_printf)(NULL, 4, "Using Libiio v0.x -> v1.x compatibility layer\n");
+
+	return;
+
+err_dlclose:
+	iio_dlclose(ctx->lib);
+err_free_ctx:
+	free(ctx);
+err_set_errno:
+	errno = -err;
+}
+
+static void compat_lib_exit(void)
+{
+	iio_dlclose(compat_lib_ptr->lib);
+	free(compat_lib_ptr);
+}
+
+#if defined(_MSC_BUILD)
+#pragma section(".CRT$XCU", read)
+#define __CONSTRUCTOR(f, p) \
+  static void f(void); \
+  __declspec(allocate(".CRT$XCU")) void (*f##_)(void) = f; \
+  __pragma(comment(linker,"/include:" p #f "_")) \
+  static void f(void)
+#ifdef _WIN64
+#define _CONSTRUCTOR(f) __CONSTRUCTOR(f, "")
+#else
+#define _CONSTRUCTOR(f) __CONSTRUCTOR(f, "_")
+#endif
+#elif defined(__GNUC__)
+#define _CONSTRUCTOR(f) static void __attribute__((constructor)) f(void)
+#else
+#define _CONSTRUCTOR(f) static void f(void)
```

Comment:
I might be missing something but how is compat_lib supposed to be initialized in this case?

Reply:
Right. I guess it would make more sense to use an `#error` and only enable the compat library with supported compilers.

Reply:
Agree...